Add extra product attributes

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

This PR adds some of the extra product attributes like GTIN, MPN, etc. when syncing them to Google Merchant Center. It adds a new tab to the product edit page and allows the custom attributes to be entered by the user via a form. It also checks for integrations by providing filters to modify the attribute's "acceptable values" and the actual value.

Part of #16.

### Testing instructions
1. Create a new product or edit an existing one
2. There should be a new tab called "Google listings and ads" on the product data section of the edit product page
3. There should be a field to enter MPN and GTIN for the product
4. You can select one of the following for these attributes
        - `Default`: This means that the value will be set to the default value specified on a global/category level (not yet implemented)
        - __Values from integrations__: For now we have only one integration for the Yoast WooCommerce Seo extension, so if you have that installed there will be a value called "From Yoast WooCommerce SEO" displayed there
        - `Enter a custom value`: If this is selected a new text box will be displayed under the current attribute to allow the user to enter their own custom value for this product
5. Choose any of the values and update/publish the product
6. Wait until the product gets synced to Google (you can check the action scheduler job or just sync it manually via the Connection Test page)
7. Check the synced product details on Google Merchant Center and make sure that the attributes you set for them are visible in there
8. Repeat the testing steps for variable products and this time set the attributes per each variation

### Additional information
- We do a simple validation on the attributes on our end. For example, GTIN length is validated but we don't validate its correctness (e.g. check digit validation).
- To test the Yoast WooCommerce SEO integration without installing it (since it's not free), you could set a global constant in `wp-config` called `WPSEO_WOO_VERSION` (with any value) and then create a new metadata for your product in the `wp_postmeta` table with the key `wpseo_global_identifier_values` and the following value as an example:
```
a:6:{s:5:"gtin8";s:8:"12345678";s:6:"gtin12";s:0:"";s:6:"gtin13";s:0:"";s:6:"gtin14";s:0:"";s:4:"isbn";s:0:"";s:3:"mpn";s:9:"123456789";}
```

nima-karimi committed:
Add extra product attributes

nima-karimi committed:
Merge branch 'trunk' into feature/extra-product-attributes

# Conflicts:
#	src/Internal/DependencyManagement/CoreServiceProvider.php
#	src/Product/WCProductAdapter.php

nima-karimi committed:
Set attributes when mapping WC product

nima-karimi committed:
rename filter

nima-karimi committed:
supply the woocommerce product to the filter

nima-karimi committed:
Get MPN and GTIN values from Yoast WooCommerce SEO

nima-karimi committed:
Merge branch 'trunk' into feature/extra-product-attributes

# Conflicts:
#	src/Product/ProductHelper.php

nima-karimi committed:
Don't set GTIN/MPN if values are empty

nima-karimi committed:
Return InputInterface in set methods

nima-karimi committed:
Add product factory class

nima-karimi committed:
Add separate variable for batch loop

nima-karimi committed:
Set identifier exists if GTIN or MPN is set

nima-karimi committed:
Get parent's meta data if product is variation

In YoastWooCommerceSeo

nima-karimi committed:
Pass form_name variable to inputs instead of form object

nima-karimi committed:
Don't return attribute object if value is null/empty

nima-karimi committed:
Add attribute form for each variation on edit product page

nima-karimi committed:
Add a default value for attributes with select options

This will allow to remove the attribute if they have previously set an option and rely on the default value provided by the extension. The default will be later on set in the plugin settings.

JPry reviewed (changes requested):
This is looking really good so far, and I like how much thought you've given to all of this. I've gotten some testing done of the functionality, and I think the only issue I encountered was what I resolved with #490.

nima-karimi committed:
Merge branch 'trunk' into feature/extra-product-attributes

nima-karimi committed:
Use `esc_attr` to sanitize form id

Also use `selected` WP function for the `select` inputs

nima-karimi committed:
validate all inputs implement the InputInterface

nima-karimi committed:
fix phpdoc

nima-karimi committed:
remove comment

nima-karimi committed:
move all views to a separate root dir

nima-karimi committed:
use `selected` WP function for select box

nima-karimi committed:
add __toString method for attributes

nima-karimi committed:
change the attributes form name

nima-karimi committed:
fix the custom attribute values not saving

nima-karimi committed:
remove unnecessary comment

nima-karimi committed:
create interface for integrations

nima-karimi committed:
Refactor the Assets classes

Use inheritance instead of traits and also allow assets to load only on certain conditions (specified by calling a closure set in the `enqueue_condition` property of the asset).

nima-karimi committed:
Use attribute description on input help tooltip

nima-karimi committed:
Refactor Assets classes

nima-karimi committed:
check for `should_register_tasks`

nima-karimi committed:
Merge pull request #511 from woocommerce/feature/refactor-assets-classes

Refactor the Assets classes

nima-karimi committed:
refactor the inputs and add js files to handle SelectWithTextInput

nima-karimi committed:
fix SelectWithTextInput toggle code

nima-karimi committed:
Refactor forms and inputs

Forms can now have children and inputs are now essentially forms with extra attributes.

nima-karimi committed:
Merge branch 'trunk' into feature/extra-product-attributes

# Conflicts:
#	src/Admin/Admin.php
#	src/Assets/ScriptHelper.php

nima-karimi committed:
Fix merge conflict

nima-karimi committed:
fix js lint errors

nima-karimi committed:
Put GLA attributes for variations in a togglable meta box

nima-karimi committed:
Merge branch 'trunk' into feature/extra-product-attributes

nima-karimi committed:
remove unused import

nima-karimi committed:
Set default attribute value to null

nima-karimi committed:
Use AbstractAttributesForm and a helper method to add attributes

nima-karimi committed:
make `get_value_options` method static

nima-karimi committed:
Use a non-abstract attributes form

nima-karimi committed:
Remove attribute on update if value is null or empty string

nima-karimi committed:
Merge branch 'trunk' into feature/extra-product-attributes

# Conflicts:
#	src/Product/ProductHelper.php

mikkamp reviewed (approved):
The code looks good, although I did leave some comments on parts we might want to revise. I'm still going to approve the PR as I think we should address any of the non minor points in follow up issues (otherwise we'll never get through this PR).

Another point we might want to create an issue for:
Save meta as `yoast_seo`, when selecting "From Yoast WooCommerce SEO"
I personally think we shouldn't be able to set this per product / variation at all. If they have gone through the trouble of entering this data in Yoast then they will just want one flip of the switch to get the data from Yoast. I understand we don't have a global way of setting attributes yet, so we can move this to a separate issue so it doesn't block the PR.

We also need to create an issue still for which attributes can be edited in the parent variable product.

nima-karimi committed:
Fix typo
